# goal definition
int8 type								# type of action
int8 predfinedPoseCode
int32 planID							# plan id for searching exist plan from oracle, to avoid duplicate planning work, not implemented
cwru_ariac/Part sourcePart				# part to pick
cwru_ariac/Part targetPart				# fake part message to place the part
float64[] jointsValue					# new joint value to be published
float64 timeout						 	# time limit for perform any move action, timeout <=0 means no limit

# type list
int8 NONE						= 0	 	# unused
int8 CONVEYOR_FETCH					= 9     # specialized to grab a part from conveyor
int8 PICK						= 10	# input: planID, sourcePart (from), timeout; output: success, errorCode, robotState
int8 PLACE						= 11	# input: planID, targetPart (to), timeout; output: success, errorCode, robotState
int8 MOVE						= 12	# input: planID, sourcePart (from), targetPart (to), timeout; output: success, errorCode, robotState
int8 TO_HOME					= 13	# input: timeout; output: success, errorCode, robotState
int8 SET_JOINT_VALUE			= 15	# input: jointsValue, timeout; output: success, errorCode, robotState
int8 GET_ROBOT_STATE			= 16	# input: (none); output: success, robotState
int8 GRASP						= 17	# input: timeout; output: success, errorCode
int8 RELEASE					= 18	# input: timeout; output: success, errorCode
int8 IS_ATTACHED				= 19	# input: (none); output: success (true means attached)
int8 TO_PREDEFINED_POSE         = 20    # input: predfinedPoseCode, timeout; output: success, errorCode

# predefined move codes
int8 AGV1_HOVER_POSE                    =101
int8 AGV1_CRUISE_POSE                   =102
int8 AGV2_HOVER_POSE                   =103
int8 AGV2_CRUISE_POSE                   =04
int8 BIN8_HOVER_POSE                   =104
int8 BIN8_RETRACT_POSE                   =105
int8 BIN8_CRUISE_POSE                   =106
int8 BIN7_HOVER_POSE                   =107
int8 BIN7_RETRACT_POSE                   =108
int8 BIN7_CRUISE_POSE                   =109
int8 BIN6_HOVER_POSE                   =110
int8 BIN6_RETRACT_POSE                   =111
int8 BIN6_CRUISE_POSE                   =112
int8 BIN5_HOVER_POSE                   =113
int8 BIN5_RETRACT_POSE                   =114
int8 BIN5_CRUISE_POSE                   =115
int8 BIN4_HOVER_POSE                   =116
int8 BIN4_RETRACT_POSE                   =117
int8 BIN4_CRUISE_POSE                   =118
int8 BIN3_HOVER_POSE                   =119
int8 BIN3_RETRACT_POSE                   =120
int8 BIN3_CRUISE_POSE                   =121
int8 BIN2_HOVER_POSE                   =122
int8 BIN2_RETRACT_POSE                   =123
int8 BIN2_CRUISE_POSE                   =124
int8 BIN1_HOVER_POSE                   =125
int8 BIN1_RETRACT_POSE                   =126
int8 BIN1_CRUISE_POSE                   =127

---
# result definition
bool success							# success or bool result, if success, errorCode will be ignored
int8 errorCode							# reason of fault, specified below
cwru_ariac/RobotState robotState        # robot state after finish execute current command

# ERROR CODE
int8 NO_ERROR			= 0				# default
int8 CANCELLED		    = 100			# goal canceled
int8 WRONG_PARAMETER	= 101			# type not in list or Part message have error
int8 TIMEOUT			= 102			# out of time limit
int8 UNREACHABLE		= 103           # pose unreachable
int8 GRIPPER_FAULT		= 104			# when gripper failed to attach or detach object
int8 COLLISION			= 105			# when robot hits anything
int8 PART_DROPPED		= 106			# part dropped in the middle, return immediately with RobotState, a replanning will be performed
---
# feedback
cwru_ariac/RobotState robotState        # current robot state when performing any action
